contracts/V4Router.sol

```diff
@@ -0,0 +1,174 @@
+                (PoolKey memory poolKey, bool oneForZero) =
+                    _getPoolAndSwapDirection(params.path[i - 1], params.currencyOut);
+                uint128 amountIn = uint128(
+                    _swapExactPrivate(
+                        poolKey,
+                        !oneForZero,
+                        -int256(int128(params.amountOut)),
+                        0,
+                        msgSender,
+                        i == 1,
+                        i == params.path.length
+                    )
+                );
+
+                params.amountOut = amountIn;
+                params.currencyOut = params.path[i - 1].tradeCurrency;
+            }
+            if (params.amountOut > params.amountInMaximum) revert TooMuchRequested();
+        }
+    }
+
+    function _swapExactPrivate(
+        PoolKey memory poolKey,
+        bool zeroForOne,
+        int256 amountSpecified,
+        uint160 sqrtPriceLimitX96,
+        address msgSender,
+        bool settle,
+        bool take
+    ) private returns (int128 reciprocalAmount) {
+        BalanceDelta delta = poolManager.swap(
+            poolKey,
+            IPoolManager.SwapParams(
+                zeroForOne,
+                amountSpecified,
+                sqrtPriceLimitX96 == 0
+                    ? (zeroForOne ? TickMath.MIN_SQRT_RATIO + 1 : TickMath.MAX_SQRT_RATIO - 1)
+                    : sqrtPriceLimitX96
+            ),
+            bytes("")
+        );
+
+        if (zeroForOne) {
+            reciprocalAmount = amountSpecified > 0 ? delta.amount1() : delta.amount0();
+            if (settle) _payAndSettle(poolKey.currency0, msgSender, delta.amount0());
+            if (take) poolManager.take(poolKey.currency1, msgSender, uint128(-delta.amount1()));
+        } else {
+            reciprocalAmount = amountSpecified > 0 ? delta.amount0() : delta.amount1();
+            if (settle) _payAndSettle(poolKey.currency1, msgSender, delta.amount1());
+            if (take) poolManager.take(poolKey.currency0, msgSender, uint128(-delta.amount0()));
+        }
+    }
+
+    function _getPoolAndSwapDirection(PathKey memory params, Currency currencyIn)
+        private
+        pure
+        returns (PoolKey memory poolKey, bool zeroForOne)
+    {
+        (Currency currency0, Currency currency1) =
+            currencyIn < params.tradeCurrency ? (currencyIn, params.tradeCurrency) : (params.tradeCurrency, currencyIn);
+
+        zeroForOne = currencyIn == currency0;
+        poolKey = PoolKey(currency0, currency1, params.fee, params.tickSpacing, params.hooks);
+    }
+
+    function _payAndSettle(Currency currency, address msgSender, int128 settleAmount) private {
+        _pay(Currency.unwrap(currency), msgSender, address(poolManager), uint256(uint128(settleAmount)));
+        poolManager.settle(currency);
+    }
+
+    function _pay(address token, address payer, address recipient, uint256 amount) internal virtual;
```

Comment:
@hensha256 wanted to outsource this to UniversalRouter...wondering what you think of adding a generically named "pay" method which internals can call abstractly.... 🤔

Reply:
So you mean like the UR would have a `pay` function with `override` in it? And then any of the contracts that it overrides can have an abstract `pay` function too?

Reply:
The `V3SwapRouter` inherits from `Permit2Payments` so it can then call `payOrPermit2Transfer` where it needs to do payments. So I guess we could change that to a `pay` function too and implement the permit2 functionality? 

I'd be interested to see how it looks in the UR repo given that currently the hierarchy is
`Permit2Payments` -> `V3SwapRouter` -> `Dispatcher`
and the `Dispatcher` doesnt have to know anything about handling payments

Reply:
yah we could leave it out of dispatcher, easiest case, have a V4SwapRouter contract in UR that inherits this and implements pay??

Reply:
> yah we could leave it out of dispatcher, easiest case, have a V4SwapRouter contract in UR that inherits this and implements pay??

yeah i think thats going to be the best solution